### PR TITLE
Category total column things

### PR DESCRIPTION
**Pull request by payten**
- Don't include Category Total columns in the fixed columns
- Group Category Totals column with grade item columns when grouping by category
- Style category total column header slightly different to the grade item column header
- Add the Category Score columns to the Show/Hide toolbar panel and hook into the fancy behaviour there
- Update the column summary to include the category score columns in counts
- Better distinguish between groups in the show/hide toggle panel

### CODE REVIEW

File: tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java [hunk 1]
```diff
@@ -49,7 +50,8 @@
 	public GradebookUiSettings() {
 		//defaults. Note there is no default for assignmentSortOrder as that requires an assignmentId which will differ between gradebooks
 		this.categoriesEnabled = false;
-		this.assignmentVisibility = new HashMap<Long, Boolean>();
+		this.assignmentVisibility = new ConcurrentHashMap<Long, Boolean>();
+		this.categoryScoreVisiblity = new ConcurrentHashMap<String, Boolean>();
 		this.nameSortOrder = GbStudentNameSortOrder.LAST_NAME;
```
Comment (payten) [inline tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java:55]: Hm.. Is there a potential concurrency issue here?  I started seeing some weird behaviour when I hit "Hide All" as it fires off an ajax post for all the grade items to update the same Hash.

I'm wondering if we need to make access or setting of this session object a little more robust to concurrent requests?

Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java:55]: The use of ConcurrentHashMap immediately raised red flags for me. Clicking 'hide all' just toggles all checkboxes to be clicked and hence fires their onUpdate method correct? I haven't applied this yet, what do you see, seems ok from here.

Perhaps hideAll/showAll could handle it a bit differently and not call each in sequence but rather just one method to update the map all at once?

**Comment by payten** [comment]
Thanks @steveswinsburg, I've rolled out the `ConcurrentHashMap` - that was my misguided attempt at making the `GradebookUiSettings` theadsafe.  

Is the Session threadsafe?  My fear was that we regularly write the `GBNG_UI_SETTINGS` value to the session and there's no guarantee that the user will have the latest or we whallop what another thread has set.

**Comment by steveswinsburg** [comment]
Session isn't threadsafe itself, but the setAttribute method is as it is atomic, however if we do getAttribute, process it, setAttribute then that is not guaranteed (multiple operations mean another thread could potentially intercept and change stuff). So we can either have a look at how we use this and maybe refactor so we don't need to get the value first and just set it, or we can mark the getUiSettings method as synchronized. However if we do that we should move the UiSettings stuff into its own static class first, otherwise the entire GradebookPage might become synchronised which would be bad.

Are you able to reliably reproduce the issue you are seeing? The issue is because it's asynchronous and so firing off a bunch of requests when that hide all/show all is clicked. I'd like to observe this, I haven't been able to reproduce. Was the issue when you clicked hide/show all that some were marked and some were left behind? 

The other alternative is that we make that call a bit differently and rather than mark them all as checked which fires its own request for each, we have a different handler for them all in bulk somehow.

**Comment by payten** [comment]
I haven't been able to replicate it (of course!), so it's more of a feeling ;)  After reading some of the Wicket doco [1] it seems that the user session is already synchronized as the whole page state is stored on the session.   So I think my worries are unfounded and I must've just been seeing a bug of my own doing!

If Wicket does synchronize the session restricting to one thread per user session, it may be worth batching the Ajax posts for show all/hide all as you suggest.   Currently the UI doesn't wait for those POST results so the user won't be left waiting around if those requests are queued and take a while.  So perhaps we'll be ok?!  Maybe something to keep an eye on.

[1] https://cwiki.apache.org/confluence/display/WICKET/Thread+Safety

**Comment by steveswinsburg** [comment]
Yeah we'll keep an eye on it. I've also been doing some reading on this and if we did do it we could probably synchronize on the GbUiSettings class itself so that anything else accessing it is blocked until that operation is complete (ie if a user manages to sort two assignments at the same time). We would end up needing the ConcurrentHashMap after all ;)

Re that link, I wonder how that plays out with AJAX requests. Though the last point is probably pertinent though "When beginning request processing one of the first things wicket does is synchronize on the user's session so that only one request thread can process a request belonging to the same session user session."

If you do see something weird we can probably do something about it.

File: tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java [hunk 2]
```diff
@@ -39,6 +39,7 @@
 	private boolean categoriesEnabled;
 
 	private Map<Long, Boolean> assignmentVisibility;
+	private Map<String, Boolean> categoryScoreVisiblity;
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java]: typo! :lollipop:

Comment (payten) [inline tool/src/java/org/sakaiproject/gradebookng/tool/model/GradebookUiSettings.java]: Thanks!

**Pull request merged**